sympy/matrices/expressions/matexpr.py

```diff
@@ -821,6 +778,133 @@ def __nonzero__(self):
 def matrix_symbols(expr):
     return [sym for sym in expr.free_symbols if sym.is_Matrix]
 
+
+class _LeftRightArgs(object):
+    def __init__(self, first=S.One, second=S.One, first_T=S.One, second_T=S.One):
+        self.first = first
+        self.second = second
+        self.first_T = first_T
+        self.second_T = second_T
+        self.trace = False
+    def __repr__(self):
+        return "_LeftRightArgs(first=%s[%s], second=%s[%s], first_T=%s[%s], second_T=%s[%s], trace=%s)" % (
+                self.first, self.first.shape,
+                self.second, self.second.shape,
+                self.first_T, self.first_T.shape,
+                self.second_T, self.second_T.shape,
+                self.trace,
+        )
+    def transpose(self):
+        return type(self)(self.first_T, self.second_T, self.first, self.second)
+    def __hash__(self):
+        return hash((self.first, self.second, self.first_T, self.second_T))
+    def __eq__(self, other):
+        if not isinstance(other, _LeftRightArgs):
+            return False
+        return (self.first == other.first) and (self.second == other.second) and (self.first_T == other.first_T) and (self.second_T == other.second_T)
+
+
+def _matrix_derivative(expr, x):
+    from .trace import Trace
+    from .transpose import Transpose
+    from sympy import (Identity, ZeroMatrix)
+
+    if isinstance(expr, MatrixSymbol):
```

Comment:
Can we move this code into the respective _eval_derivative methods of each class?

Reply:
I guess each class should have a `_eval_derivative_lines` or something like that and the base class MatrixExpr would implement `_eval_derivative` would be your `matrix_derivative` function.

Reply:
> Can we move this code into the respective _eval_derivative methods of each class?

Theoretically yes, I have developed the code in one block for simplicity. I would first expand the tests a lot to make sure that everything works.

I need very long expressions with intertwines expression trees to check whether the visit of the expression tree is correct.

Reply:
You can generate random expressions, but is there any way to check that the result is correct? I guess you can check it against the matrix calculus online calculator.

Reply:
Yes, there is that website. And we can still check by going the explicit-index way with Kronecker deltas.